### ISSUE TITLE
Crash when adding new field to geojson via feautre attribute table

### ISSUE THREAD

**Issue by magnus-m** [qgis/QGIS#51262]
## User Feedback

Adding a new field to a geosjon file via the table attribute viewer, then trying to save the edit.
Note that it still will work when adding fields via _layer properties_ --> _Fields_

![image](https://user-images.githubusercontent.com/19188046/208428426-004f3d88-3eb7-47bd-bd50-0dac4c1a2985.png)

## Report Details

**Python Stack Trace**
```
Windows fatal exception: access violation

Current thread 0x00089600 (most recent call first):
<no Python frame>
```

**Stack Trace**

```

QVariant::~QVariant :
QgsSQLiteExpressionCompiler::operator= :
QgsSQLiteExpressionCompiler::operator= :
QgsSQLiteExpressionCompiler::operator= :
QgsAbstractFeatureIterator::nextFeature :
QgsVectorLayerFeatureIterator::fetchFeature :
QgsAbstractFeatureIterator::nextFeature :
QgsCachedFeatureWriterIterator::fetchFeature :
QgsAbstractFeatureIterator::nextFeature :
QgsAttributeTableModel::loadLayer :
QObject::qt_static_metacall :
QObject::qt_static_metacall :
QgsVectorLayer::emitDataChanged :
QObject::qt_static_metacall :
QgsPrecalculatedTextMetrics::count :
QgsVectorLayer::commitChanges :
QgsProject::commitChanges :
QgisApp::saveVectorLayerEdits :
QgisApp::saveEdits :
QgsAttributeTableDialog::mActionSaveEdits_triggered :
QObject::qt_static_metacall :
QAction::activate :
QAbstractButton::click :
QAbstractButton::mouseReleaseEvent :
QToolButton::mouseReleaseEvent :
QWidget::event :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QApplicationPrivate::sendMouseEvent :
QSizePolicy::QSizePolicy :
QSizePolicy::QSizePolicy :
QApplicationPrivate::notify_helper :
QApplication::notify :
QgsApplication::notify :
QCoreApplication::notifyInternal2 :
QGuiApplicationPrivate::processMouseEvent :
QWindowSystemInterface::sendWindowSystemEvents :
QEventDispatcherWin32::processEvents :
qt_plugin_query_metadata :
QEventLoop::exec :
QCoreApplication::exec :
main :
BaseThreadInitThunk :
RtlUserThreadStart :
```

### Steps to reproduce the issue

1. open QGIS
2. create scratch layer
3. create one feature
4. save scratch layer as geojson
5. open table attributes 
6. turn on editing for geojson layer
7. create new field
8. save edits

### Versions

**QGIS Info**
QGIS Version: 3.28.1-Firenze
QGIS code revision: fde3b8fbb8c
Compiled against Qt: 5.15.3
Running against Qt: 5.15.3
Compiled against GDAL: 3.6.0
Running against GDAL: 3.6.0

**System Info**
CPU Type: x86_64
Kernel Type: winnt
Kernel Version: 10.0.22000

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

**Comment by roya0045** [comment]
Geojson were always wonky to edit/amend since they are feature based and not a table, but I never had it crash.. just ignore or delete changes..

**Comment by magnus-m** [comment]
I updated the **Steps to reproduce the issue** to include how to create a json file. The content is also included here:

`test.geojson`
```json
{
    "type": "FeatureCollection",
    "name": "json_test",
    "crs": {
        "type": "name",
        "properties": {
            "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
        }
    },
    "features": [
        {
            "type": "Feature",
            "properties": {
                "id": null
            },
            "geometry": {
                "type": "Point",
                "coordinates": [
                    -1.172080772607551,
                    0.450395083406497
                ]
            }
        }
    ]
}
```

**Comment by magnus-m** [comment]
> Geojson were always wonky to edit/amend since they are feature based and not a table, but I never had it crash.. just ignore or delete changes..

I think it would be better for the user experience to remove the option to add a field from the attribute table when working with geojson, than to have it crash.

**Comment by elpaso** [comment]
@magnus-m agreed: this is a bug and should be fixed.

**Comment by elpaso** [comment]
Probably introduced by https://github.com/qgis/QGIS/pull/46750 , the stacktrace points to `QgsVectorLayerEditBuffer::commitChangesCheckAttributesModifications`.

CC @domi4484

**Comment by elpaso** [comment]
@domi4484 can you have a look at this issue?

**Comment by domi4484** [comment]
@elpaso yes i can take a look tomorrow

**Comment by domi4484** [comment]
It crashes at different places for me, this one is quite interesting:

- `attributeData[idx] = getFeatureAttribute( fet.get(), idx );` in `QgsOgrFeatureIterator::readFeature`

`idx ` is out of range. It makes me think the crash is more related to the fact that the new column is not really added in the beginning. But only when some values are added because of the nature of geojson.

If I add some values to the new created column before saving, it doesn't crash anymore.

**Comment by roya0045** [comment]
> It crashes at different places for me, this one is quite interesting:
> 
> * `attributeData[idx] = getFeatureAttribute( fet.get(), idx );` in `QgsOgrFeatureIterator::readFeature`
> 
> `idx ` is out of range. It makes me think the crash is more related to the fact that the new column is not really added in the beginning. But only when some values are added because of the nature of geojson.
> 
> If I add some values to the new created column before saving, it doesn't crash anymore.

That's not surprising, something similar happen if you configure a memory layer, don't add any feature and then make it permanent as a geojson, as no feature are written QGIS can't guess the geometry and no fields are present.

**Comment by elpaso** [comment]
@domi4484 IIRC the crash was coming from the call `QgsVectorLayerEditBuffer::commitChangesCheckAttributesModifications` which calls `L->updateFields()`  at a point in time when the new field was added to the layer but not yet to the provider.

**Comment by domi4484** [comment]
I have looked a little more into this and the problem is in `QgsOgrProvider::leaveUpdateMode`. There, after the call to `reloadData()`, the just added field disappear (see screenshot).
In `reloadData()` a `dataChanged` signal is emitted, which causes a reload in the `QgsAttributeTableModel`. At this point the feature iterator tries to get values for the new field which is (for a short time) not known to the provider.

Adding the `Qt::QueuedConnection` flag to postpone the execution of `QgsAttributeTableModel::loadLayer` solves the crash. But it solves just the symptoms, the underlying issue remain.

![image](https://user-images.githubusercontent.com/9881900/211276595-d8c82dc6-9e9d-4af7-9272-9fd857dc9204.png)

**Comment by domi4484** [comment]
Fixed in #51494 (I can't close the issue)
